### ISSUE TITLE
Perf: Fuse FFF steps

### ISSUE THREAD

**Issue by ClashLuke** [sap-ient-ai/FFF#2]
This line:
https://github.com/sap-ient-ai/FFF/blob/25c1a03228e7862cd5a98d5745a681dd1353b05f/FFF/fff.py#L45

and the equivalent for the output computes a MxKx1 matrix multiplication. Depending on the backend hardware (A100/H100, TPUv2/3/4), it'd beneficial to unroll steps of the loop. You could do this by `gather`-ing all required nodes for N steps such as `x, 2x+1, 2x+2, 4x+3, 4x+4, 4x+5, 4x+6`, following
https://github.com/sap-ient-ai/FFF/blob/25c1a03228e7862cd5a98d5745a681dd1353b05f/FFF/fff.py#L57
and matmul-ing with all of them. As an A100 needs to launch at least a MxKx16 matrix multiplication, this should not increase computational cost. Further unrolling may be beneficial in increasing arithmetic intensity.

**Comment by p-i-** [comment]
Yes, if I understand you right, gathering indices and then doing a single matmul projection to y would be faster.
The original authors did this.
I golfed the original as I wanted to create a reference implementation that makes it ultra-easy to grok what's going on. (hmm maybe I shouldn't have used einsum...).

Maybe there should be a separate reference impl, and a performant impl.

It's a challenge to consider where/whether to optimize.
I've knocked up a trite JAX version which should optimize out the python `for` loop.
Bojan is working on a CUDA impl.
Meanwhile I'm finding a family of designs that train better (at some expense, so there's question of where to position ones self on the computation-to-trainingAccuracy curve).
